Add support for read consistency levels

Blackmoomba opened the pull request:
Adds optional param in Cursor.execute which allows the caller to set a read consistency level.

See https://rqlite.io/docs/api/read-consistency/

otoolep commented:
Thank you for your contribution. Before I merge this I would like to see a couple of unit tests added to https://github.com/rqlite/pyrqlite/blob/master/src/test/test_dbapi.py#L97 (The connection test class).

Blackmoomba commented:
That's fair. I'll update after adding tests.

otoolep commented:
E.g.
```python
    def test_CheckExecuteParamListConsistency(self):
        self.cu.execute("insert into test(name) values ('foo')")
        self.cu.execute("select name from test where name= foo", consistency='strong')
        row = self.cu.fetchone()
        self.assertEqual(row[0], "foo")
```

otoolep commented:
Or maybe:
```python
    def test_CheckExecuteParamListConsistency(self):
        self.cu.execute("insert into test(name) values ('foo')")
        for c in ['strong', 'none', 'weak']:
            self.cu.execute("select name from test where name= foo", consistency=c)
            row = self.cu.fetchone()
            self.assertEqual(row[0], "foo")
```
The test is not particularly strong, but at least checks that setting the param doesn't get wrong results. I would do 'strong' first, that way you can be sure when 'none' is used and the database is accessed directly without any Leader check, that the database does contain previous writes.

otoolep commented on src/pyrqlite/cursors.py:
```diff
@@ -163,8 +163,10 @@ def execute(self, operation, parameters=None, queue=False, wait=False):
+            path = "/db/query?"
+            if consistency in ("none", "weak", "strong"):
+                path = path + "level=" + consistency
+            payload = self._request("GET", path + _urlencode({'q': operation}))
```
Actually, I'm not sure this will work.

If consistency is explicitly set, then you need a URL of the form:

`/db/query?q=SELECT&consistency=...`

I do not see this handled correctly in the code. We need to use requests properly now, see https://requests.readthedocs.io/en/latest/user/quickstart/#passing-parameters-in-urls

It's going to be something like this:

```python
params = {'q': operation}
if consistency in ("none", "weak", "strong"):
    params['consistency'] = consistency
```

otoolep replied:
Or do: `_urlencode({'q': operation, 'consistency': c}))` and default `consistency` to `weak`.

Blackmoomba replied:
Ah, glaring miss on my part. I see how the query is built now, I'll fix this as well when I add tests.

Blackmoomba replied:
I disagree about setting `consistency` to `weak` by default here. I think it's probably more appropriate to not set a consistency level param at all if one is not expressly provided by the user. From the rqlite docs, it seems that the default is `weak` if one isn't provided. If that default were to change for any reason in the future in the server, then it would also need to change here.

With that said, I'm happy to commit to setting that default if that's your preference.

otoolep replied:
Yeah, that's fine. I think your position is just as correct TBH.

otoolep replied:
So I suggest we create the query params dict, and set `q`. And only if the client sets `consistency` explicitly should the code also add a `consistency` key, so no consistency is set unless explicitly set by the client. That's what I might do anyway. 

By the way, to your point that things can change in the database itself, that argues against even checking if any supplied consistency is one of `none`, `weak`, or `strong`. While I doubt it will happen, rqlite mihgt support a new level in the future, and this client library would need to change again. Instead you just take whatever string is set for consistency (if it's not None) and just set it.

Blackmoomba replied:
That's valid. We'll rely on rqlite's error messages here if the user provides an invalid consistency level.

Blackmoomba committed:
Add support for read consistency levels

Adds optional param in Cursor.execute which allows the caller to set a read consistency level.

otoolep commented:
Thanks for your contribution.
